File: 567-permutation-in-string/567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        
       if(s1.size() > s2.size()) return false;
       
       int s1map[26]={0};
       int s2map[26]={0};
        
        for(int i = 0 ; i < s1.size(); i ++ ){
            int x = s1[i]-'a';
            int y = s2[i]-'a';
            s1map[x]++;
            s2map[y]++;
        }
        int count = 0 ;
        for(int i = 0; i < 26; i ++){
            if(s1map[i]==s2map[i]) count ++ ;
        }
        
        for(int i = 0 ; i < s2.size()- s1.size(); i ++ ){
            int r = s2[i+s1.size()] - 'a';
            int l = s2[i] - 'a';
            if(count == 26) return true;
            s2map[r]++;
            if(s2map[r] == s1map[r]){
                count ++;
            }else if(s2map[r] == s1map[r]+1){
                count --;
            }
            s2map[l]--;
            if(s2map[l] == s1map[l]){
                count ++;
            }else if(s2map[l] == s1map[l]-1){
                count --;
            }  
                
        }
        
        return (count == 26);
        

    }
};

File: 567-permutation-in-string/567-permutation-in-string.swift
class Solution {
    
    func isContainsAnyChar(_ mp: [Character: Int]) -> Bool{
        let firstLetter = Unicode.Scalar("a")
        let lastLetter = Unicode.Scalar("z")
        var count = 0
        for val in firstLetter.value...lastLetter.value{
            if let char = Unicode.Scalar(val){
                // print(char)
                if mp[Character(char), default:0] == 0{
                    count += 1
                }
            }
        }
        return (count != 26)
    }
    
    func checkInclusion(_ s1: String, _ s2: String) -> Bool {
        if s2.count < s1.count {
            return false
        }
        var windowStart = 0, windowEnd = 0
        
        var mp = [Character: Int]()
        
        var s1Array = Array(s1), s2Array = Array(s2)
        
        for index in 0..<s1Array.count{
            mp[s2Array[index], default: 0] -= 1
            mp[s1Array[index], default: 0] += 1 
        }
        
        windowEnd = s1Array.count
        while( windowEnd < s2Array.count){
            if (!isContainsAnyChar(mp)){
                return true
            }else{
                
                mp[s2Array[windowStart], default :0] += 1
                mp[s2Array[windowEnd], default :0] -= 1
                
                windowStart += 1
                windowEnd += 1
            }
        }
        
        if (!isContainsAnyChar(mp)) {
            return true
        }
        else {
            return false
        }
    }
}